src/apps/competitions/unpackers/base_unpacker.py

```diff
@@ -0,0 +1,294 @@
+            'id',
+            'primary_index',
+            'title',
+            'key',
+            'columns' : {
+                'id',
+                'computation',
+                'computation_indexes',
+                'title',
+                'key',
+                'sorting',
+                'index',
+            },
+        }
+        """
+        raise NotImplementedError
+
+    def _save_tasks(self):
+        for index, task in self.competition['tasks'].items():
+            if isinstance(task, str):
+                try:
+                    task = Task.objects.get(key=task)
+                except Task.DoesNotExist:
+                    raise CompetitionUnpackingException(f'Task with key {task} does not exist')
+                self.competition['tasks'][index] = task
+            else:
+                for file_type in ['ingestion_program', 'input_data', 'scoring_program', 'reference_data']:
+                    try:
+                        task_file_data = task[file_type]
+                    except KeyError:
+                        # file type not here, moving on
+                        continue
+                    key = self._get_data_key(**task_file_data)
+                    task[file_type] = key
+                serializer = TaskSerializer(
+                    data=task,
+                )
+                serializer.is_valid(raise_exception=True)
+                new_task = serializer.save()
+                self.created_tasks.append(new_task)
+                self.competition['tasks'][index] = new_task
+
+    def _save_solutions(self):
+        for solution in self.competition['solutions']:
+            if 'key' in solution:
+                try:
+                    s = Solution.objects.get(key=solution['key'])
+                except Solution.DoesNotExist:
+                    raise CompetitionUnpackingException(f'Could not find solution with key: {solution["key"]}')
+                for task_index in solution['tasks']:
+                    s.add(self.competition['tasks'][task_index])
+            else:
+                solution['tasks'] = [self.competition['tasks'][index].key for index in solution['tasks']]
+                solution['data'] = self._get_data_key(**solution, file_type='solution')
+                serializer = SolutionSerializer(data=solution)
+                serializer.is_valid(raise_exception=True)
+                new_solution = serializer.save()
+                self.created_solutions.append(new_solution)
+
+    def _save_competition(self):
+        for phase in self.competition['phases']:
+            phase['tasks'] = [self.competition['tasks'][index].key for index in phase['tasks']]
+        serializer = CompetitionSerializer(
+            data=self.competition,
+            context={'created_by': self.creator}
+        )
+        serializer.is_valid(raise_exception=True)
+        return serializer.save()
+
+    def _clean(self):
+        # TODO Figure out how to delete from storage, too
```

Comment:
can remove todo with the `delete()` override you have, I think?

Reply:
Good catch.
✔️